kubernetes/cache/kube_cache.go

```diff
@@ -729,21 +729,27 @@ func (c *kubeCache) GetServices(namespace string, selectorLabels map[string]stri
 	var services []*core_v1.Service
 	var err error
 	if namespace == metav1.NamespaceAll {
-		services, err = c.getCacheLister(namespace).serviceLister.List(labels.Set(selectorLabels).AsSelector())
+		services, err = c.getCacheLister(namespace).serviceLister.List(labels.Everything())
 	} else {
-		services, err = c.getCacheLister(namespace).serviceLister.Services(namespace).List(labels.Set(selectorLabels).AsSelector())
+		services, err = c.getCacheLister(namespace).serviceLister.Services(namespace).List(labels.Everything())
 	}
 	if err != nil {
 		return nil, err
 	}
 	log.Tracef("[Kiali Cache] Get [resource: Service] for [namespace: %s] = %d", namespace, len(services))
 
+	selector := labels.Set(selectorLabels)
 	retServices := []core_v1.Service{}
 	for _, service := range services {
-		// Do not modify what is returned by the lister since that is shared and will cause data races.
-		svc := service.DeepCopy()
-		svc.Kind = kubernetes.ServiceType
-		retServices = append(retServices, *svc)
+		svcSelector := labels.Set(service.Spec.Selector).AsSelector()
+		// selector match is done after listing all services, similar to registry reading
+		// empty selector is loading all services, or match the service selector
+		if selector.AsSelector().Empty() || (!svcSelector.Empty() && svcSelector.Matches(selector)) {
+			// Do not modify what is returned by the lister since that is shared and will cause data races.
+			svc := service.DeepCopy()
+			svc.Kind = kubernetes.ServiceType
+			retServices = append(retServices, *svc)
+		}
```

Comment:
This is basically a "field selector" now that selects the `service.Spec.Selector` which is not what you'd expect when you call `kubeCache.GetServices(...)`. It might have weird effects elsewhere. I think it'd be better to keep the `kubeCache.GetServices` method as is and have a separate function for filtering a slice of `[]core_v1.Service` by their `spec.Selector` matching a `map[string]string`.

Reply:
`kubeCache.GetServices` method will stay working wrongly for cases when `selectorLabels` is provided.